src/seq.rs

```diff
@@ -236,6 +285,125 @@ pub trait IteratorRandom: Iterator + Sized {
+    ///
+    /// let choices = ['a', 'b', 'c'];
+    /// let weights = [2,   1,   1];
+    /// let mut rng = thread_rng();
+    /// // 50% chance to print 'a', 25% chance to print 'b', 25% chance to print 'c'
+    /// println!("{}", weights.choose_index_weighted(&mut rng).map(|ix| choices[ix]).unwrap());
+    /// ```
+    /// [`into_weighted_index_distribution`]: trait.IntoIteratorRandom.html#method.into_weighted_index_distribution
+    /// [`Vec`]: https://doc.rust-lang.org/std/vec/struct.Vec.html
+    fn choose_index_weighted<R, X>(self, rng: &mut R) -> Option<usize>
+        where Self::IntoIter: Clone,
+              R: Rng + ?Sized,
+              X: SampleUniform +
+                 for<'a> ::core::ops::AddAssign<&'a X> +
+                 ::core::cmp::PartialOrd<X> +
+                 Default,
+              Self::Item: SampleBorrow<X> {
+        let iter = self.into_iter();
+        let mut total_weight: X = Default::default();
+        let zero: X = Default::default();
+        for w in iter.clone() {
+            assert!(*w.borrow() >= zero, "Weight must be >= zero");
+            total_weight += w.borrow();
+        }
+
+        if *total_weight.borrow() == zero {
+            return None;
+        }
+
+        let chosen_weight = rng.gen_range(zero, total_weight);
+        let mut cumulative_weight = <X as Default>::default();
+
+        for (i, weight) in iter.enumerate() {
+            cumulative_weight += weight.borrow();
+            if cumulative_weight > chosen_weight {
+                return Some(i);
+            }
+        }
+
+        panic!("Weights changed during cloning");
+    }
+
+    /// Returns a [`Distribution`] which when sampled from, returns the index
+    /// of a random element from this iterator where the chance of a given
+    /// item being picked is proportional to the value of the item.
+    ///
+    /// # Panics
+    ///
+    /// If a value in the iterator is `< 0`, or if all the values in the
+    /// iterator returns `0`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use rand::prelude::*;
+    ///
+    /// let choices = ['a', 'b', 'c'];
+    /// let weights = [2,   1,   1];
+    /// let dist = weights.into_weighted_index_distribution();
+    /// let mut rng = thread_rng();
+    /// for _ in 0..100 {
+    ///     // 50% chance to print 'a', 25% chance to print 'b', 25% chance to print 'c'
+    ///     println!("{}", choices[dist.sample(&mut rng)]);
+    /// }
+    /// ```
+    #[cfg(feature = "alloc")]
+    fn into_weighted_index_distribution<X>(self) -> WeightedIndex<X>
+        where X: SampleUniform +
+                 for<'a> ::core::ops::AddAssign<&'a X> +
+                 ::core::cmp::PartialOrd<X> +
+                 Clone +
+                 Default,
+              Self::Item: SampleBorrow<X> {
+        let mut iter = self.into_iter();
+        let mut total_weight: X = iter.next()
+                                      .expect("Can't create Distribution for empty set of weights")
```

Comment:
I don't think we should panic in this case. Maybe return some type of error instead.